### PR TITLE
✨ Update FixedLayer on Element Hidden Toggle

### LINKED ISSUE #16554: Fixed position banner is sometimes rendered out of the viewport
## What's the issue?

Our banner that is fixed to the bottom of the screen (`position:fixed, bottom: 0px`) is sometimes rendered fully or halfway out of the viewport. Both issues can be "fixed" by scrolling to the bottom of the page and swiping down, as if you were doing a pull-to-refresh motion on the bottom of the screen. 

img_4233

Here are a few videos that demonstrate the issue to help in case you are unable to reproduce:

- [Banner rendering partially out of the viewport](https://drive.google.com/file/d/1eJoqn_v_yM1H_oMETp7mjZOYFqdO0jOI/view?usp=sharing)
- [Banner rendering fully out of the viewport](https://drive.google.com/file/d/1wxHpRE0TFhD-4uFDDkX5QA5sXpTDqlKf/view?usp=sharing)
- [Swiping through AMP pages that all have partially obstructed banner](https://drive.google.com/file/d/1X8jTC18lPMKJ_Y6PDjd8XKB_X4K7JCa1/view?usp=sharing)

## How do we reproduce the issue?
_This issue is most easily reproducible on an actual phone, rather than by emulating a device in Chrome dev tools._ 

1. Make sure you are logged out of Pinterest by going to www.pinterest.com/logout
2. Visit https://www.google.com/amp/s/www.pinterest.com/amp/pin/27866091429714425/ to see halfway out example
3. Visit https://www.google.com/amp/s/www.pinterest.com/amp/pin/5840674496841567/ to see fully out example

Expected result: Black banner at the bottom is fully visible
Actual result: Black banner is halfway or fully hidden and obstructed by the view port

## What browsers are affected?

Chrome v67.0.3396.87 and Firefox v12.1 (10941) are affected. Safari does not appear to be. Have not tried other versions of Chrome/Firefox.

Reproducible on iPhone X and iPhone 6s. Have not tried other devices.

## Which AMP version is affected?

Unable to tell at this time, since I can only repro on a phone.

### LINKED ISSUE #16933: AMP viewer blocking top sticky banner on iOS
## What's the issue?
The AMP Viewer sometimes blocks sticky banner at the top of the page, leaving the area of the page that was meant to have the banner on it with just whitespace. 

screen shot 2018-07-19 at 1 34 47 pm

[Screen recording of behavior](https://drive.google.com/file/d/1KAKBMs_70fkPcVmhDAG436eU3my4yyXA/view?usp=sharing)

## How do we reproduce the issue?
1. Search for something that will produce a Pinterest AMP result, like "Pinterest home decor"
2. Click to open Pin page and notice app install banner on top

Note, if you see the app install banner on the bottom of the screen, please clear your cache and try again (this feature is still in experiment).

## What browsers are affected?

iOS Safari, iOS Chrome

## Which AMP version is affected?

Unsure.

### LINKED ISSUE #17475: FixedLayer: Transfer elements on [hidden] toggle
Problem: Fixed elements that are initially hidden may have incorrect DOM ordering WRT fixed elements that are not initially hidden. 

Proposed solution: Use MutationObserver to listen for tracked, fixed elements that are un-hidden dynamically by removing `hidden` attribute (e.g. via amp-bind or toggle/show actions). Transfer those elements to the fixed layer when they're un-hidden.

Context: `b/111906735`

### PR DESCRIPTION
**Pull request by jridgewell**
Now when an element is toggled (hidden or un-hidden), the `FixedLayer` will run an update.

Fixes #17475
Fixes #16554
Fixes #16933
Depends on #18788

### CODE REVIEW

File: src/service/fixed-layer.js [hunk 1]
```diff
@@ -137,6 +153,65 @@ export class FixedLayer {
...
+  unobserveHiddenMutations_() {
+    this.clearMutationObserver_();
+    const mo = this.mutationObserver_;
+    if (mo) {
+      mo.disconnect();
+    }
+  }
+
+  /**
+   * Clears the mutation observer and it's pass queue.
```
Comment (choumx) [inline src/service/fixed-layer.js]: Nit: its

File: src/service/fixed-layer.js [hunk 2]
```diff
@@ -137,6 +153,65 @@ export class FixedLayer {
...
+  clearMutationObserver_() {
+    this.updatePass_.cancel();
+    const mo = this.mutationObserver_;
+    if (mo) {
+      mo.takeRecords();
+    }
+  }
+
+  /**
+   * @return {?MutationObserver}
```
Comment (choumx) [inline src/service/fixed-layer.js]: `!MutationObserver`

File: src/service/fixed-layer.js [hunk 3]
```diff
@@ -137,6 +153,65 @@ export class FixedLayer {
...
+
+    const mo = new this.ampdoc.win.MutationObserver(mutations => {
+      if (this.updatePass_.isPending()) {
+        return;
+      }
+
+      for (let i = 0; i < mutations.length; i++) {
+        const mutation = mutations[i];
+        if (mutation.attributeName === 'hidden') {
+          this.updatePass_.schedule(16);
```
Comment (choumx) [inline src/service/fixed-layer.js]: Explain magic number in a comment?

**Comment by aghassemi** [comment]
@jridgewell @choumx would this make it to Canary tomorrow of 10/30 release?

File: src/service/fixed-layer.js [hunk 4]
```diff
@@ -138,6 +153,66 @@ export class FixedLayer {
...
+    const mo = new this.ampdoc.win.MutationObserver(mutations => {
+      if (this.updatePass_.isPending()) {
+        return;
+      }
+
+      for (let i = 0; i < mutations.length; i++) {
+        const mutation = mutations[i];
+        if (mutation.attributeName === 'hidden') {
+          // Wait one animation frame so that other mutations may arrive.
+          this.updatePass_.schedule(16);
```
Comment (choumx) [inline src/service/fixed-layer.js]: For a follow-up: bit of a waste to run the whole `update()` on a single element's `hidden` attribute change, no? Can we constrain the operation to just the mutated elements?

**Comment by jridgewell** [comment]
> would this make it to Canary tomorrow of 10/30 release?

Yah, I'll wait for this to land before cutting canary. I'm on duty this week, so it's lucky.

**Pull request merged**